support fixed timesteps in direct trajectory optimization

RussTedrake opened the pull request:
(formulations with time as a decision variable are rarely convex)

Also resolves a TODO with the placeholder vars.

N.B. - the test coverage of the trajectory optimization classes is very weak... and it makes changes painful.  This is the first of a train of PRs to start cleaning that up.  I also have a PR queued up to allow DirectCollocation to use fixed timesteps, but it was a big change so I've split it up.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6850)
<!-- Reviewable:end -->

RussTedrake commented:
+@ericcousineau-tri for feature review, please.

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-:-KreS_q1-S6nckOugKov:b7kye1q)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, pending some minor comments and interface questions. Thanks!

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 20 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krf_2UaVxuHjuZ8x4CV:-Krf_2UbW4fPsYXcVgY3:b-uccs5n) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_collocation.cc#L20)):*
> ```cpp
>           system->get_num_total_inputs(),
>           context.get_continuous_state()->size(), num_time_samples,
>           trajectory_time_lower_bound / (num_time_samples - 1),
> ```

BTW Rather than using `trajectory_time_{upper,lower}_bound`, could this instead be named `duration_{upper,lower}_bound` to make the relation to `min_timestep` and `max_timestep` a little easier to infer?
(At an initial glance, I first thought `time_{upper,lower}_bound` implied start and end time.)

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfgK2nolW0Uqi8iv-q:-KrfgK2o4FM7qRbTNLT2:b-s0r9h7) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_collocation.cc#L21)):*
> ```cpp
>           context.get_continuous_state()->size(), num_time_samples,
>           trajectory_time_lower_bound / (num_time_samples - 1),
>           trajectory_time_upper_bound / (num_time_samples - 1)),
> ```

BTW As a follow-up, it feels odd to see the signature for `DirectTrajectoryOptimization` operate on timestep, while `DirectCollocation` operates on total duration.

It seems like it'd involve some refactoring, but should one match to the other, perhaps as a separate issue? (It seems like maybe making `DirectTrajectoryOptimization` match `DirectCollocation` would be the easiest from a user's perspective.)

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfmTGjvOk9JjVi8dvW:-KrfmTGjvOk9JjVi8dvX:b-g1vrth) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_collocation.cc#L27)):*
> ```cpp
>   DRAKE_DEMAND(context.has_only_continuous_state());
> 
>   // TODO(russt):  This should NOT be set automatically
> ```

BTW Grammar, comma or semi-colon after "automatically"?

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfmdUOnlmY6uOqcT91:-KrfmdUOnlmY6uOqcT92:b-44pvr8) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_collocation.cc#L28)):*
> ```cpp
> 
>   // TODO(russt):  This should NOT be set automatically
>   // once it is removed, the proper constraint to be added will be
> ```

BTW Can I ask what the envisioned use case would be for not automatically setting constraints on the step times?
Would you still need supply a minimum / maximum timestep w/ number of samples at construction?

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.cc, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrflydOBTqJRqoU4_hz:-KrfdG68XPFnEF1FEH-N:b-scekb) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.cc#L109)):*
> ```cpp
>                                                      double upper_bound) {
>   DRAKE_THROW_UNLESS(timesteps_are_decision_variables_);
>   AddLinearConstraint(VectorXd::Ones(h_vars_.size()), lower_bound, upper_bound,
> ```

BTW nit In `ConstructPlaceholderVariableSubstitution`, `h_vars_.cast<Expression>().sum()` was used. Could this be made into a `GetSymbolicDuration()` helper, and be used here for consistency?

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.cc, line 133 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfdugOIL2dckJ1WGDm:-KrfdugOIL2dckJ1WGDn:b-ez45u0) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.cc#L133)):*
> ```cpp
>     DRAKE_DEMAND(start_time <= end_time);
>     SetInitialGuess(h_vars_, VectorXd::Constant(h_vars_.size(),
>                                                 (end_time - start_time) / N_));
> ```

BTW Could `(end_time - start_time)  / N_` be called `h_init` for minor clarification?

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.cc, line 135 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krfh5baIzk1G-WYRE5T:-Krfh5baIzk1G-WYRE5U:b-uwkeqo) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.cc#L135)):*
> ```cpp
>                                                 (end_time - start_time) / N_));
>   }
>   const double timestep = (end_time - start_time) / N_;
> ```

BTW super minor nit For consistency, should this be called `h`?

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 203 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krfl44Y_iSEjbghaMKM:-Krfl44ZpUJ62ip7_rq8:by6b9r1) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L203)):*
> ```objc
>   /// @param traj_init_u Initial guess for trajectory for control
>   /// input. The number of rows for each segment in @p traj_init_u must
>   /// be equal to num_inputs (the first param of the constructor).  If
> ```

BTW minor nit Two spaces at the beginning of the sentence. (One space used elsewhere)

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 211 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfhNL8wCW8HKYCY2pF:-KrfhNL8wCW8HKYCY2pG:b-vhhlf7) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L211)):*
> ```objc
>   /// empty, then a default small non-zero initial value is used instead.
>   ///
>   /// If time steps are decision variables, then the initial guess for
> ```

Should there be documentation on the logic for the start and end time if both `traj_init_u` and `traj_init_x` are supplied with different times?

It seems like it will take the common time interval between the two, and effectively shift it to zero for the returned trajectory, e.g., if `t_u = [0.2, 5]`, and `t_x = [0.5, 1]`, and if the timesteps do not change, then my assumption is that the returned trajectory would be `t_sol = [0, 0.5]`, with an effective offset of `0.5` to the initial guess time intervals. (Which may not be a huge issue, if they're just seed values.)

Alternatively, if using purely fixed timesteps, then it looks like it would just use clamped values of `t` for `traj_u` and `traj_x`.

As an alternative, should this demand that the intervals match, and be equivalent to the initial guess for the time interval?

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 252 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krfhjo1o9eHnGzb09xi:-Krfhjo1o9eHnGzb09xj:b-1jukku) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L252)):*
> ```objc
>   /// @param num_time_samples Number of time samples.
>   /// @param minimum_timestep Minimum spacing between sample times.
>   /// @param maximum_timestep Minimum spacing between sample times.
> ```

Typo, "Maximum spacing..."

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 257 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfleXRcacXY8cUSIpP:-KrfleXRcacXY8cUSIpQ:b-ek1x1u) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L257)):*
> ```objc
>                                double maximum_timestep);
> 
>   /// Returns a vector containing the elapsed time at each knot point.
> ```

BTW Should this documentation indicate that the knot points are from the solution?

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfeORYm64upkdajSOt:-KrfeORZSPeKrYvnEsdm:b-16sgty) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L53)):*
> ```cpp
> GTEST_TEST(TrajectoryOptimizationTest, FixedTimestepTest) {
>   MyDirectTrajOpt prog(1, 2, 2,
>                        0.1);  // 1 input, 2 states, 2 timesteps, dt=0.1.
> ```

BTW minor nit As an alternative to annotating the call site with comments, would it be possible to define these as constants to then pass?

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfefCpYfVtVwR3gUAH:-KrfefCqyrvGaEqKQHb5:b7haey7) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L79)):*
> ```cpp
> 
> GTEST_TEST(TrajectoryOptimizationTest, VariableTimestepTest) {
>   MyDirectTrajOpt prog(1, 2, 2, 0.1,
> ```

BTW Could you add either augment the comment, or add constants, that indicate that `dt_min = dt_max = 0.1`?

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 112 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krff7qBcvvZRdURfgrv:-Krff7qBcvvZRdURfgrw:bglgoxb) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L112)):*
> ```cpp
>   prog.SetInitialGuess(prog.timestep(1), Vector1d(.2));
> 
>   EXPECT_NE(prog.GetSolution(prog.timestep(0)),
> ```

Inspecting the code for `MathematicalProgram`, I think `EXPECT_NE` only works here because `x_values_` from `GetSolution` is initialized to `NaN`.
Was the intent to check that they are initialized to `NaN`?

Or should this instead use `initial_guess()` and/or should we add a corresponding `GetInitialGuess(vars)` method to check the initial values?

(That being said, it seems that this test in general may be redundant if `SetInitialGuess` is already tested in `MathematicalProgram`, since you are using the same decision variables.)

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfiS7FZP9FsvfQZvSa:-KrfiS7FZP9FsvfQZvSb:b-fu3sb3) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L131)):*
> ```cpp
>   prog.SetInitialGuess(prog.timestep(1), Vector1d(.2));
> 
>   EXPECT_NE(prog.GetSolution(prog.h_vars()).array().sum(), .5);
> ```

Same as above, was the intent here to check for `NaN` behavior? If not, should it check `EXPECT_EQ(0.3, total_time)`?

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 175 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfisBtIMfgMK9l2dfm:-KrfisBtIMfgMK9l2dfn:bsc6hw7) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L175)):*
> ```cpp
>   const std::vector<double> times{0.0, t_init_in};
> 
>   const PiecewisePolynomialType inputs_u(u_vec, times);
> ```

BTW `inputs_u` and `states_x` seems a tad redundant.
Could this perhaps be called `traj_u` and `traj_x`, or `traj_input` and `traj_state`? (with or without `init_` prefix?)

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 279 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfZq8whk6IG5Fk1eYW:-KrfZq8xU-1VA-h6zeQL:b9scmsq) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L279)):*
> ```cpp
>   MyDirectTrajOpt prog(kNumInputs, kNumStates, kNumTimeSamples, 0, 25);
> 
>   // Adding a placeholder variable directly to the program should fail.
> ```

Since the `EXPECT_THROW` passes and #5623 now passes, should this comment be removed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-:-KrfoUAysOW8guzcrgYO:b-hf6bdg)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 0 of 8 files reviewed at latest revision, 17 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 20 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krf_2UaVxuHjuZ8x4CV:-Krg5KM8c4Ri7iSLWSFi:b9wq8uo) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_collocation.cc#L20)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Rather than using `trajectory_time_{upper,lower}_bound`, could this instead be named `duration_{upper,lower}_bound` to make the relation to `min_timestep` and `max_timestep` a little easier to infer?
(At an initial glance, I first thought `time_{upper,lower}_bound` implied start and end time.)
</blockquote></details>

that's the old interface.  i'm changing it in my next PR.

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfgK2nolW0Uqi8iv-q:-Krg5OxXk3LbYPimqWw4:bk51gnk) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_collocation.cc#L21)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW As a follow-up, it feels odd to see the signature for `DirectTrajectoryOptimization` operate on timestep, while `DirectCollocation` operates on total duration.

It seems like it'd involve some refactoring, but should one match to the other, perhaps as a separate issue? (It seems like maybe making `DirectTrajectoryOptimization` match `DirectCollocation` would be the easiest from a user's perspective.)
</blockquote></details>

also coming in the next PR

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfmTGjvOk9JjVi8dvW:-Krg5hFUgQjxnd_uOTA_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_collocation.cc#L27)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Grammar, comma or semi-colon after "automatically"?
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfmdUOnlmY6uOqcT91:-Krg5iumh6cPlcyWzMEo:b-dqvsps) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_collocation.cc#L28)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Can I ask what the envisioned use case would be for not automatically setting constraints on the step times?
Would you still need supply a minimum / maximum timestep w/ number of samples at construction?
</blockquote></details>

again, cleanup of the dircol is coming in the next PR.  

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.cc, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrflydOBTqJRqoU4_hz:-Krg5tIhratwdZKJEY_N:bda8oc0) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.cc#L109)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW nit In `ConstructPlaceholderVariableSubstitution`, `h_vars_.cast<Expression>().sum()` was used. Could this be made into a `GetSymbolicDuration()` helper, and be used here for consistency?
</blockquote></details>

the two use cases are a little different so far... this one is simple enough that I can trivially avoid the overhead of going through symbolic expression (the other one is inserting the time into potentially more complicated expressions).  will add it as soon as we have multiple common use cases.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.cc, line 133 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfdugOIL2dckJ1WGDm:-Krg6tFs6Ryg8hAfsgSW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.cc#L133)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Could `(end_time - start_time)  / N_` be called `h_init` for minor clarification?
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.cc, line 135 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krfh5baIzk1G-WYRE5T:-Krg6t_GIuuTiLvrXexD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.cc#L135)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW super minor nit For consistency, should this be called `h`?
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 203 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krfl44Y_iSEjbghaMKM:-Krg72-ChQK8jUSQiKC_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L203)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW minor nit Two spaces at the beginning of the sentence. (One space used elsewhere)
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 211 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfhNL8wCW8HKYCY2pF:-Krg8BMi8_Bp1jJFjSN4:b-13gait) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L211)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Should there be documentation on the logic for the start and end time if both `traj_init_u` and `traj_init_x` are supplied with different times?

It seems like it will take the common time interval between the two, and effectively shift it to zero for the returned trajectory, e.g., if `t_u = [0.2, 5]`, and `t_x = [0.5, 1]`, and if the timesteps do not change, then my assumption is that the returned trajectory would be `t_sol = [0, 0.5]`, with an effective offset of `0.5` to the initial guess time intervals. (Which may not be a huge issue, if they're just seed values.)

Alternatively, if using purely fixed timesteps, then it looks like it would just use clamped values of `t` for `traj_u` and `traj_x`.

As an alternative, should this demand that the intervals match, and be equivalent to the initial guess for the time interval?
</blockquote></details>

Done.  Actually went overkill (changed the API to clean up).  Also added proper test coverage for this complex API.  

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 252 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krfhjo1o9eHnGzb09xi:-Krgm6uZlfGMzj7_EmEl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L252)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Typo, "Maximum spacing..."
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 257 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfleXRcacXY8cUSIpP:-KrgmKnb3ESL9pAxn4nV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L257)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Should this documentation indicate that the knot points are from the solution?
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfeORYm64upkdajSOt:-KrgmPPQqKxppiFpz3L4:bv43u7o) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L53)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW minor nit As an alternative to annotating the call site with comments, would it be possible to define these as constants to then pass?
</blockquote></details>

It's way more lines of code for something simple that's repeated a bunch of times in the test suite.  I'm hoping to keep it as is.

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfefCpYfVtVwR3gUAH:-KrgmoHevDS1BJvVAQuD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L79)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Could you add either augment the comment, or add constants, that indicate that `dt_min = dt_max = 0.1`?
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 112 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krff7qBcvvZRdURfgrv:-KrgnI-Q_POFmZdi6xbG:bpqra17) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L112)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Inspecting the code for `MathematicalProgram`, I think `EXPECT_NE` only works here because `x_values_` from `GetSolution` is initialized to `NaN`.
Was the intent to check that they are initialized to `NaN`?

Or should this instead use `initial_guess()` and/or should we add a corresponding `GetInitialGuess(vars)` method to check the initial values?

(That being said, it seems that this test in general may be redundant if `SetInitialGuess` is already tested in `MathematicalProgram`, since you are using the same decision variables.)
</blockquote></details>

Good catch.  fixed.

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfiS7FZP9FsvfQZvSa:-KrgnK80mjd6sobY7oMi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L131)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Same as above, was the intent here to check for `NaN` behavior? If not, should it check `EXPECT_EQ(0.3, total_time)`?
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 175 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfisBtIMfgMK9l2dfm:-KrgnMeRKEQrCVdPi3LU:bs50fp) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L175)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW `inputs_u` and `states_x` seems a tad redundant.
Could this perhaps be called `traj_u` and `traj_x`, or `traj_input` and `traj_state`? (with or without `init_` prefix?)
</blockquote></details>

this is old test code, which I will soon delete.

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 279 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfZq8whk6IG5Fk1eYW:-KrgnWq2mDDnvYiap95_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L279)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Since the `EXPECT_THROW` passes and #5623 now passes, should this comment be removed?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6850)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 211 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrfhNL8wCW8HKYCY2pF:-KrgsI-DSN-iKHPbVqr0:b-ucfknh) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L211)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.  Actually went overkill (changed the API to clean up).  Also added proper test coverage for this complex API.  
</blockquote></details>

Thanks!

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 112 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krff7qBcvvZRdURfgrv:-KrgsWSzw6GRxPgbyc6I:b-eogaor) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L112)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Good catch.  fixed.
</blockquote></details>

BTW Both here and below (total time), would it be possible to check the exact values, rather than checking inequality?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6850)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 7 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc, line 112 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krff7qBcvvZRdURfgrv:-KrhPJBIFbEZppG_ed_7:b-c1ns2n) ([raw file](https://github.com/robotlocomotion/drake/blob/557e6a27fb1a6812f97431a1a28150034da50c6f/drake/systems/trajectory_optimization/test/trajectory_optimization_test.cc#L112)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Both here and below (total time), would it be possible to check the exact values, rather than checking inequality?
</blockquote></details>

i could.  IMHO, the existence of those NE tests are only valuable in that they demonstrate that the state changes (from NE to EQ) by calling Solve().  So my goal is to highlight the contrast, as opposed to the specific values...?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6850)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
+@soonho-tri for platform review please (i fear i've missed the wednesday cycle)

---

Review status: 7 of 8 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-:-KrhhMZNtJR_-jGUgM62:b-r9jz6r)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> aside a few BTWs.
Please feel free to address them in the next PRs if you want.

---

Reviewed 7 of 8 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.cc, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krkrnfsn44z028aJu6Q:-Krkrnfsn44z028aJu6R:b-d527oy) ([raw file](https://github.com/robotlocomotion/drake/blob/294eaa3ce8d0ca4c6d8f19f6b0ac6d07a0e94f20/drake/systems/trajectory_optimization/direct_trajectory_optimization.cc#L27)):*
> ```cpp
> namespace {
> 
> solvers::VectorXDecisionVariable MakeNamedVariables(const std::string prefix,
> ```

BTW (resolved), @jwnimmer-tri wanted to have this helper function before. @jwnimmer-tri do you think it's worth having it in `solvers/decision_variables.h` and/or `symbolic_variable.h`? If so, I'll open another PR for that.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.cc, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrkshTT32wKbwyYHZFZ:-KrkshTT32wKbwyYHZF_:b-hy2haf) ([raw file](https://github.com/robotlocomotion/drake/blob/294eaa3ce8d0ca4c6d8f19f6b0ac6d07a0e94f20/drake/systems/trajectory_optimization/direct_trajectory_optimization.cc#L27)):*
> ```cpp
> namespace {
> 
> solvers::VectorXDecisionVariable MakeNamedVariables(const std::string prefix,
> ```

BTW, `const std::string prefix` -> `const std::string& prefix`.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 171 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrkoPOWoAd3Sq4KYnH0:-KrkoPOWoAd3Sq4KYnH1:b-sivcxd) ([raw file](https://github.com/robotlocomotion/drake/blob/294eaa3ce8d0ca4c6d8f19f6b0ac6d07a0e94f20/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L171)):*
> ```objc
>   /// for all time intervals.
>   /// @param lower_bound  A scalar double lower bound.
>   /// @param upper_bound  A scalar double upper bound.
> ```

BTW, I think it's worth adding a doxygen documentation such as
```c++
/// @throws if timesteps_are_decision_variables_ is false.
```
for the methods which have `DRAKE_THROW_UNLESS(timesteps_are_decision_variables_)` in the body.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 174 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrkncRHrX80SkBW-fYM:-KrkncRI86_E5sMzj1fs:bscfox3) ([raw file](https://github.com/robotlocomotion/drake/blob/294eaa3ce8d0ca4c6d8f19f6b0ac6d07a0e94f20/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L174)):*
> ```objc
>   void AddTimeIntervalBounds(double lower_bound, double upper_bound);
> 
>   /// Add constraints to enforce that all timesteps have equal duration.
> ```

BTW, META nit, `s/Add/Adds/`.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 291 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krkr3ylsXSpemRo9ADL:-Krkr3ylsXSpemRo9ADM:b-256x1h) ([raw file](https://github.com/robotlocomotion/drake/blob/294eaa3ce8d0ca4c6d8f19f6b0ac6d07a0e94f20/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L291)):*
> ```objc
>     return fixed_timestep_;
>   }
>   const solvers::VectorXDecisionVariable& h_vars() const { return h_vars_; }
> ```

BTW, I think it's good to have the following here:
```c++
    DRAKE_THROW_UNLESS(timesteps_are_decision_variables_);
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-:-KrkvD5t0AeZ5uItIJOh:baqwl7r)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.cc, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krkrnfsn44z028aJu6Q:-Krl2cT3ijW5vUVW2HAs:b-lhlu2o) ([raw file](https://github.com/robotlocomotion/drake/blob/294eaa3ce8d0ca4c6d8f19f6b0ac6d07a0e94f20/drake/systems/trajectory_optimization/direct_trajectory_optimization.cc#L27)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW (resolved), @jwnimmer-tri wanted to have this helper function before. @jwnimmer-tri do you think it's worth having it in `solvers/decision_variables.h` and/or `symbolic_variable.h`? If so, I'll open another PR for that.
</blockquote></details>

yes.  definitely.  thanks.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.cc, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrkshTT32wKbwyYHZFZ:-Krl2f3yhB0BsVhRwBsz:b-e91p1k) ([raw file](https://github.com/robotlocomotion/drake/blob/294eaa3ce8d0ca4c6d8f19f6b0ac6d07a0e94f20/drake/systems/trajectory_optimization/direct_trajectory_optimization.cc#L27)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `const std::string prefix` -> `const std::string& prefix`.
</blockquote></details>

Done (in follow-on PR).

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 171 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrkoPOWoAd3Sq4KYnH0:-Krl3-l3f5kNyqcyzMDF:b8aagef) ([raw file](https://github.com/robotlocomotion/drake/blob/294eaa3ce8d0ca4c6d8f19f6b0ac6d07a0e94f20/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L171)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, I think it's worth adding a doxygen documentation such as
```c++
/// @throws if timesteps_are_decision_variables_ is false.
```
for the methods which have `DRAKE_THROW_UNLESS(timesteps_are_decision_variables_)` in the body.
</blockquote></details>

Done (in the follow-on PR).

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 174 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-KrkncRHrX80SkBW-fYM:-Krl4In8K0tsLsxB1Esv:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/294eaa3ce8d0ca4c6d8f19f6b0ac6d07a0e94f20/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L174)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, META nit, `s/Add/Adds/`.
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_trajectory_optimization.h, line 291 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6850#-Krkr3ylsXSpemRo9ADL:-Krl3GW3H13mnkDx2ejB:b1ybrd3) ([raw file](https://github.com/robotlocomotion/drake/blob/294eaa3ce8d0ca4c6d8f19f6b0ac6d07a0e94f20/drake/systems/trajectory_optimization/direct_trajectory_optimization.h#L291)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, I think it's good to have the following here:
```c++
    DRAKE_THROW_UNLESS(timesteps_are_decision_variables_);
```
</blockquote></details>

i intentionally didn't add one here because it's protected and because h_vars_ is actually defined (its the empty vector).  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6850)*
<!-- Sent from Reviewable.io -->
